### PR TITLE
Gun adapter

### PR DESCRIPTION
**Pull request by alex-strizhakov**
Adapter for gun - https://github.com/ninenines/gun

Added custom option - max_body.

### CODE REVIEW

File: lib/tesla/adapter/gun.ex [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+      stream
+    end
+
+    defp open_stream(pid, method, url, headers, body, false),
+      do: :gun.request(pid, method, url, headers, body)
+
+    defp read_response(pid, stream, opts) do
+      receive do
+        msg ->
+          case msg do
```
Comment (rinpatch) [inline lib/tesla/adapter/gun.ex]: Can't you just write clauses in the `recieve` block directly?

File: lib/tesla/adapter/gun.ex [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+          {:error, "read response timeout"}
+      end
+    end
+
+    defp read_body(pid, stream, mref, opts, acc \\ "") do
+      limit = opts[:max_body]
+
+      receive do
+        msg ->
+          case msg do
```
Comment (rinpatch) [inline lib/tesla/adapter/gun.ex]: Can't you just write clauses in the recieve block directly?

File: lib/tesla/adapter/gun.ex [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+                read_body(pid, stream, mref, opts, acc <> part)
+              else
+                {:error, "body too large"}
+              end
+
+            msg ->
+              IO.inspect(msg)
+              IO.inspect("unsupported message received in read body.")
+              msg
+          end
```
Comment (rinpatch) [inline lib/tesla/adapter/gun.ex]: Should handle [:gun_error](https://ninenines.eu/docs/en/gun/1.3/guide/http/#_processing_responses)

File: lib/tesla/adapter/gun.ex [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    mix deps.compile tesla
+    ```
+    ### Example usage
+    ```
+    # set globally in config/config.exs
+    config :tesla, :adapter, Pleroma.Tesla.Adapter.Gun
+    # set per module
+    defmodule MyClient do
+      use Tesla
+      adapter Pleroma.Tesla.Adapter.Gun
```
Comment (teamon) [inline lib/tesla/adapter/gun.ex]:
```suggestion
      adapter Tesla.Adapter.Gun
```

File: lib/tesla/adapter/gun.ex [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    Remember to add `{:gun, "~> 1.3"}` to dependencies
+    Also, you need to recompile tesla after adding `:gun` dependency:
+    ```
+    mix deps.clean tesla
+    mix deps.compile tesla
+    ```
+    ### Example usage
+    ```
+    # set globally in config/config.exs
+    config :tesla, :adapter, Pleroma.Tesla.Adapter.Gun
```
Comment (teamon) [inline lib/tesla/adapter/gun.ex]:
```suggestion
    config :tesla, :adapter, Tesla.Adapter.Gun
```

File: lib/tesla/adapter/gun.ex [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    end
+    ```
+
+    ### Options:
+
+    * `connect_timeout` - Connection timeout.
+    * `http_opts` - Options specific to the HTTP protocol.
+    * `http2_opts` -  Options specific to the HTTP/2 protocol.
+    * `protocols` - Ordered list of preferred protocols. Defaults: [http2, http] - for :tls, [http] - for :tcp.
+    * `retry` - Number of times Gun will try to reconnect on failure before giving up. Default: 5
```
Comment (teamon) [inline lib/tesla/adapter/gun.ex]: There is `Tesla.Middleware.Retry` middleware that works with all adapters, exposing this option might be confusing (which one should be used?)

File: mix.exs [hunk 7]
```diff
@@ -69,6 +70,7 @@ defmodule Tesla.Mixfile do
       # other
       {:fuse, "~> 2.4", optional: true},
       {:telemetry, "~> 0.3", optional: true},
+      {:cowlib, "~> 1.0.2", override: true, optional: true},
```
Comment (teamon) [inline mix.exs]: Why is this needed?

File: test/tesla/adapter/gun_test.exs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  use Tesla.AdapterCase.SSL
+
+  test "timeout option" do
+    request = %Env{
+      method: :get,
+      url: "#{@http}/delay/2"
+    }
+
+    assert {:error, "read response timeout"} = Tesla.Adapter.Gun.call(request, timeout: 1_000)
+  end
```
Comment (teamon) [inline test/tesla/adapter/gun_test.exs]: While there are some issues with Tesla.Middleware.Timeout I don't think we should expose such custom options as part of adapter interface. In general adapter options should be used as last-resort case when one absolutely have to pass some options to the underlying adapter.

File: mix.exs (hunk 7 above)
Comment (alex-strizhakov) [reply mix.exs]: gun requires {:cowlib, "~> 2.6.0"}, that's why `override: true` was added

File: lib/tesla/adapter/gun.ex (hunk 6 above)
Comment (alex-strizhakov) [reply lib/tesla/adapter/gun.ex]: removed these options

File: test/tesla/adapter/gun_test.exs (hunk 8 above)
Comment (alex-strizhakov) [reply test/tesla/adapter/gun_test.exs]: now it uses `opts[:timeout]` option or adapter default fallback setting

File: mix.exs (hunk 7 above)
Comment (alex-strizhakov) [reply mix.exs]: removed after update `httparrot` dependency

**Comment by rinpatch** [comment]
@teamon could you give this another review please?

**Comment by teamon** [comment]
Please rebase to recent master (it includes CI fixes). Once all the tests pass I'd be happy to merge this MR 🚀

**Comment by alex-strizhakov** [comment]
@teamon can elixir 1.4.5 and otp 18 be removed?

**Comment by teamon** [comment]
I’m currently on holidays, I’ll get back to you next week.

**Comment by teamon** [comment]
Dropping support for elixir 1.4 & otp 18 would be a breaking change and would require releasing tesla 2.0. 

However, if there is no easy way to provide 1.4 support I'm ok with that, 1.4 caused few other issues already and we have elixir 1.19 already.

**Comment by alex-strizhakov** [comment]
[httparrot 1.2.0](https://github.com/edgurgel/httparrot/blob/master/mix.exs) requires `con_cache` 0.13, which can't be compiled on 1.4.5.

**Comment by teamon** [comment]
@alex-strizhakov And do we have to bump httparrot? :)

**Comment by alex-strizhakov** [comment]
@teamon without httparrot we need to have `{:cowlib, "~> 1.0.2", override: true, optional: true}` in `mix.exs`.

**Comment by teamon** [comment]
... and `gun` requires `cowlib`. 

Ok, let's drop 1.4 & 18 🔥

**Comment by alex-strizhakov** [comment]
> ... and gun requires cowlib.
Forgot about it 😅
